File: JAVA/gym.java
/*
programmers 탐욕법(Greedy) 체육복
https://programmers.co.kr/learn/courses/30/parts/12081
*/

public class gym {
    public static void main(String[] args){
        // Test Case 1
        // int n=5;
        // int[] lost = {2,4};
        // int[] reserve = {1,3,5};

        // My Test Case
        int n=10;
        int[] lost = {3,9,10};
        int[] reserve = {3,8,9};

        gym_Solution gym = new gym_Solution();
        int answer = gym.solution(n, lost, reserve);

        System.out.println(answer);
    }
}
class gym_Solution {
    public int solution(int n, int[] lost, int[] reserve) {
        for(int i=0;i<lost.length;i++)
            for(int j=0;j<reserve.length;j++)
                if(lost[i] == reserve[j]){ // 여벌이 있는데 잃어버린 학생
                    lost[i] = -1;
                    reserve[j] = -1;
                    break;
                }
        
        for(int i=0;i<lost.length;i++)
            for(int j=0;j<reserve.length;j++)
                if(lost[i]-1 == reserve[j]){ // 잃어버렸는데 여벌이 있는 학생에게 빌림
                    lost[i] = -1;
                    reserve[j] = -1;
                    break;
                }else if(lost[i]+1 == reserve[j]){ // 잃어버렸는데 여벌이 있는 학생에게 빌림
                    lost[i] = -1;
                    reserve[j] = -1;
                    break;
                }

        // 잃어버렸는데 빌리지도 못한 학생 카운트
        int count=0;
        for(int i=0;i<lost.length;i++)
            if(lost[i] != -1) count++;
        
        int answer = n - count;
        return answer;
    }
}

File: Python/Greedy/gym.py
"""
https://programmers.co.kr/learn/courses/30/parts/12244
PROGRAMMERS GREEDY GYM SUIT
"""
def solution(n, lost, reserve):
    # 여벌을 가져온 학생이 도난 당한 경우 각 list에서 제외
    for i in reserve:
        if i in lost:
            reserve.remove(i)
            lost.remove(i)
    
    # 규칙에 따라 체육복 빌려주기
    i = 0
    while True:
        # terminal condition
        if i == len(lost): break

        if lost[i]-1 in reserve:
            reserve.remove(lost[i]-1)
            lost.pop(i)
            continue
        if lost[i] in reserve:
            reserve.remove(lost[i])
            lost.pop(i)
            continue
        if lost[i]+1 in reserve:
            reserve.remove(lost[i]+1)
            lost.pop(i)
            continue
        i += 1

    answer = n - len(lost)
    return answer

def main():
    n1 = 5
    n2 = 5
    n3 = 11
    n4 = 5
    n5 = 10
    
    lost1 = [2, 4]
    lost2 = [2, 4]
    lost3 = [2, 4, 6, 8]
    lost4 = [1, 2]
    lost5 = [3, 9, 10]
    
    reserve1 = [1, 3, 5]
    reserve2 = [3]
    reserve3 = [2, 5, 6, 7]
    reserve4 = [2, 3]
    reserve5 = [3, 8, 9]

    result1 = 5
    result2 = 4
    result3 = 11
    result4 = 4
    result5 = 9

    if solution(n1, lost1, reserve1) == result1: print("RIGHT1")
    if solution(n2, lost2, reserve2) == result2: print("RIGHT2")
    if solution(n3, lost3, reserve3) == result3: print("RIGHT3")
    if solution(n4, lost4, reserve4) == result4: print("RIGHT4")
    if solution(n5, lost5, reserve5) == result5: print("RIGHT5")

if __name__=="__main__":
    main()